### PR TITLE
Text to speech

### PR DESCRIPTION
**Pull request by Robert-Kolmos**
Implement basic functionality of text to speech.
- Add TextToSpeechService
- Wire up text to speech for active/countdown steps.

### CODE REVIEW

**Comment by Robert-Kolmos** [comment]
I ended up using a static mapping for the special keys ("start", "end", etc). Also there is some stuff going on with a mediator live data for when to stop the service that I don't feel is very clean. If you have any suggestions for a better way to do that let me know.

File: presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+        bindCount.postValue((bindCount.value ?: 0) + 1)
+        if (LOGGER.isDebugEnabled) {
+            LOGGER.debug("onBind() called bindCount = $bindCount")
+        }
+
+        return serviceBinder
+    }
+
+    override fun onUnbind(intent: Intent?): Boolean {
+        bindCount.postValue(bindCount.value!! - 1)
```
Comment (dephillipsmichael) [inline presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt]: It's rare where you actually want to do the null-assertion operator, because you are disabling one of the best features of Kotlin.  Try this instead:

val bindCountValue = (bindCount.value ?: 1) - 1
bindCount.postValue(bindCountValue)

File: presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+        textToSpeech?.shutdown()
+        textToSpeech = null
+        cleanupMediator.removeObserver(cleanupObserver)
+        clear()
+        super.onDestroy()
+    }
+
+    override fun onInit(status: Int) {
+        if (status == TextToSpeech.SUCCESS) {
+            val languageAvailable = textToSpeech!!.isLanguageAvailable(Locale.getDefault())
```
Comment (dephillipsmichael) [inline presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt]: val launguageAvailable = textToSpeech?..isLanguageAvailable(Locale.getDefault()) ?: 0

File: presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+            // >= 0 means LANG_AVAILABLE, LANG_COUNTRY_AVAILABLE, or LANG_COUNTRY_VAR_AVAILABLE
+            if (languageAvailable >= 0) {
+                textToSpeech!!.language = Locale.getDefault()
```
Comment (dephillipsmichael) [inline presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt]: At this point you should probably do a let call to avoid the "!!"...

textToSpeech?.let {
   // here you reference textToSpeech as "it" and it won't be optional
}

File: presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+
+    /**
+     * Register the given speech map to be spoken at the times determined by the given countdown and duration.
+     * @param duration the duration of the current step
+     * @param countdown the countdown LiveData for the current step
+     * @param speechMap the map where keys are offsets into the countdown and values are instructions that should be
+     * spoken to the user.
+     */
+    fun registerSpeechesOnCountdown(duration: Duration, countdown: LiveData<Long>, speechMap: Map<String, String>) {
+        if (LOGGER.isDebugEnabled) {
```
Comment (dephillipsmichael) [inline presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt]: Should I be checking if isDebugEnabled before every log statement? 

If so, it seems odd to have to do that extra line, and we could add an extension function to use instead of .debug() something like this...

```
fun Logger.debugChecked(log: String?) {
   if (this.isDebugEnabled) {
    this.debug(log)
  }
}
```
Then you can just call LOGGER.debugChecked() everywhere that extension is imported

File: presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+
+    /**
+     * Stops speaking and removes all registered speeches.
+     */
+    fun clear() {
+        if (LOGGER.isDebugEnabled) {
+            LOGGER.debug("clear() called")
+        }
+
+        val tts = textToSpeech
```
Comment (dephillipsmichael) [inline presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt]: textToSpeech?.let {
         // Stop speaking and shut down the current text to speech.
        if (it.isSpeaking) {
           it.stop()
        }

        it.shutdown()
        textToSpeech = null
}

File: presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+     * @param text the text to speak to the user.
+     */
+    fun speakText(text: String) {
+        if (LOGGER.isDebugEnabled) {
+            LOGGER.debug("speakText() called with text \"$text\"")
+        }
+
+        val tts = textToSpeech
+        // Setting this will guarantee the text gets spoken in the case that tts isn't set up yet
+        textToSpeakOnInit = text
```
Comment (dephillipsmichael) [inline presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt]: same goes for this except you would also include the else log like this

textToSpeech?.let {
   // version check
} ?: run {
    LOGGER.warn("...")
}

File: presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+        toneG.startTone(ToneGenerator.TONE_CDMA_HIGH_L, duration)
+    }
+
+    @RequiresPermission(permission.VIBRATE)
+    fun vibrate(duration: Long = defaultVibrateAndSoundDurationMillis) {
+        if (LOGGER.isDebugEnabled) {
+            LOGGER.debug("vibrate() called with duration = $duration")
+        }
+
+        val v = getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
```
Comment (dephillipsmichael) [inline presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt]:
```
(getSystemService(Context.VIBRATOR_SERVICE) as? Vibrator)?.let {
         it.vibrate(duration)
}
```

File: presentation/src/main/java/org/sagebionetworks/research/presentation/inject/TextToSpeechModule.kt [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+@Module
+abstract class TextToSpeechModule {
+    @Multibinds
+    abstract fun specialKeyMap(): Map<String, (Duration) -> Long>
+
+    companion object {
+        @Provides
+        @IntoMap
+        @StringKey("start")
```
Comment (dephillipsmichael) [inline presentation/src/main/java/org/sagebionetworks/research/presentation/inject/TextToSpeechModule.kt]: Nice, this is a great way to allow for custom spoken keys.

File: presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt [hunk 9]
```diff
@@ -0,0 +1,437 @@
...
+
+                    override fun onError(text: String?) {
+                        _state.postValue(TextToSpeechState(ERROR, null))
+                    }
+
+                    override fun onStart(text: String?) {
+                        _state.postValue(TextToSpeechState(SPEAKING, text))
+                    }
+                })
+                if (textToSpeakOnInit != null) {
```
Comment (dephillipsmichael) [inline presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt]: `textToSpeakOnInit?.let { speakText(it) }`

File: presentation/src/main/java/org/sagebionetworks/research/presentation/inject/TextToSpeechModule.kt (hunk 8 above)
Comment (dephillipsmichael) [reply presentation/src/main/java/org/sagebionetworks/research/presentation/inject/TextToSpeechModule.kt]: Building on this map injection, we could have a map of string keys to a Unit that returns a Duration and has the StepView as a parameter.

File: presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt [hunk 10]
```diff
@@ -0,0 +1,437 @@
...
+    private val bindCount = MutableLiveData<Int>()
+    private val cleanupMediator = MediatorLiveData<Pair<Int, TextToSpeechState>>()
+    private val cleanupObserver: Observer<Pair<Int, TextToSpeechState>>
+
+    init {
+        bindCount.postValue(0)
+        _state.postValue(TextToSpeechState(IDLE, null))
+        // We observe the state so that if every client unbinds and the service stops speaking, the service
+        // stops itself.
+        cleanupMediator.postValue(Pair(0, TextToSpeechState(IDLE, null)))
```
Comment (dephillipsmichael) [inline presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt]: It's hard to follow this because it doesn't seem explained anywhere what the first val of the Pair is.

**Comment by dephillipsmichael** [comment]
No blockers from me if it works! Mostly language nits

File: presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt (hunk 4 above)
Comment (Robert-Kolmos) [reply presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt]: Oops it looks as though this check isn't needed. I can't remember where I first saw it but I'll remove it from this pr.

File: presentation/src/main/java/org/sagebionetworks/research/presentation/inject/TextToSpeechModule.kt (hunk 8 above)
Comment (Robert-Kolmos) [reply presentation/src/main/java/org/sagebionetworks/research/presentation/inject/TextToSpeechModule.kt]: I think this may be too general to be very useful. The problem I see is that if you have some sort of crazy logic based on things other than the duration of the StepView you can just process that in the step view, and then pass that map to the TTSS, and it doesn't seem like there would be a lot of overlap between different types of StepView's in this regard.

File: presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt (hunk 4 above)
Comment (liujoshua) [reply presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt]: This is mostly for when the operations to generate the String is expensive, e.g. complex String building. If you want an extension function I think it would need to take a Function<String> rather than String.
